File: src/main/java/cope/nebula/client/Nebula.java
package cope.nebula.client;

import cope.nebula.client.manager.*;
import cope.nebula.client.manager.hole.HoleManager;
import cope.nebula.client.manager.macro.MacroManager;
import cope.nebula.client.manager.relation.RelationshipManager;
import cope.nebula.util.internal.timing.Stopwatch;
import cope.nebula.util.internal.timing.TimeFormat;
import cope.nebula.util.renderer.FontUtil;
import cope.nebula.util.versioning.BuildConfig;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.lwjgl.opengl.Display;

import static net.minecraftforge.fml.common.Mod.EventHandler;
import static net.minecraftforge.fml.common.Mod.Instance;

@Mod(modid = Nebula.MODID, name = Nebula.NAME, version = Nebula.VERSION)
public class Nebula {
    public static final String NAME = "Nebula";
    public static final String VERSION = "1.0";
    public static final String HASH = BuildConfig.HASH;
    public static final String FULL_VERSION = VERSION + "-" + HASH;
    public static final String MODID = BuildConfig.NAME;

    @Instance
    private static Nebula INSTANCE;
    private static final Logger LOGGER = LogManager.getLogger(NAME);

    private ModuleManager moduleManager;
    private CommandManager commandManager;
    private MacroManager macroManager;
    private RelationshipManager relationshipManager;
    private HoleManager holeManager;

    private RotationManager rotationManager;
    private InteractionManager interactionManager;
    private HotbarManager hotbarManager;
    private ServerManager serverManager;

    @EventHandler
    public void init(FMLInitializationEvent event) {
        // Roxanne, Roxanne... All she wanna do is party all night
        LOGGER.info("Loading {} v{}...", NAME, VERSION);

        Stopwatch stopwatch = new Stopwatch();
        stopwatch.resetTime();

        FontUtil.registerCustomFonts();
        new ForgeEventManager();

        moduleManager = new ModuleManager();
        commandManager = new CommandManager();
        macroManager = new MacroManager();
        relationshipManager = new RelationshipManager();
        holeManager = new HoleManager();

        rotationManager = new RotationManager();
        interactionManager = new InteractionManager();
        hotbarManager = new HotbarManager();
        serverManager = new ServerManager();

        // Goddamn, Roxanne... Never gonna love me but it's alright
        LOGGER.info("Completed setup of {} in {}ms", NAME, stopwatch.getTime(TimeFormat.MILLISECONDS));

        Display.setTitle(NAME + " v" + FULL_VERSION);
    }

    public static Nebula getInstance() {
        return INSTANCE;
    }

    public static Logger getLogger() {
        return LOGGER;
    }

    public ModuleManager getModuleManager() {
        return moduleManager;
    }

    public CommandManager getCommandManager() {
        return commandManager;
    }

    public MacroManager getMacroManager() {
        return macroManager;
    }

    public HoleManager getHoleManager() {
        return holeManager;
    }

    public RelationshipManager getRelationshipManager() {
        return relationshipManager;
    }

    public RotationManager getRotationManager() {
        return rotationManager;
    }

    public InteractionManager getInteractionManager() {
        return interactionManager;
    }

    public HotbarManager getHotbarManager() {
        return hotbarManager;
    }

    public ServerManager getServerManager() {
        return serverManager;
    }
}

File: src/main/kotlin/wtf/nebula/client/Nebula.kt
package wtf.nebula.client

import me.bush.eventbuskotlin.Config
import me.bush.eventbuskotlin.EventBus
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.event.FMLInitializationEvent
import net.minecraftforge.fml.common.event.FMLPostInitializationEvent
import org.apache.logging.log4j.LogManager
import org.lwjgl.opengl.Display
import wtf.nebula.client.event.translate.ForgeEventListener
import wtf.nebula.client.feature.alt.Alt
import wtf.nebula.client.feature.alt.AltType
import wtf.nebula.client.manager.InventoryManager
import wtf.nebula.client.manager.RotationManager
import wtf.nebula.client.manager.server.ServerManager
import wtf.nebula.client.manager.TotemPopManager
import wtf.nebula.client.registry.RegistryContainer
import wtf.nebula.client.registry.impl.CommandRegistry
import wtf.nebula.client.registry.impl.ModuleRegistry
import wtf.nebula.util.versioning.BuildConfig

@Mod(useMetadata = true, clientSideOnly = true, modid = Nebula.ID)
class Nebula {
    companion object {
        const val NAME = "Nebula"
        const val ID = BuildConfig.NAME
        const val VERSION = BuildConfig.VERSION
        const val HASH = BuildConfig.HASH

        const val FULL = "$NAME v$VERSION/git-$HASH"

        val logger = LogManager.getLogger(NAME)
        val BUS = EventBus(Config(logger, annotationRequired = true))

        val inventoryManager = InventoryManager()
        val rotationManager = RotationManager()
        val serverManager = ServerManager()
        val totemPopManager = TotemPopManager()
    }

    @Mod.EventHandler
    fun init(event: FMLInitializationEvent) {
        logger.info("Loading $FULL")

        RegistryContainer.register(ModuleRegistry())
        RegistryContainer.register(CommandRegistry())

        BUS.subscribe(ForgeEventListener())
        BUS.subscribe(inventoryManager)
        BUS.subscribe(rotationManager)
        BUS.subscribe(serverManager)
        BUS.subscribe(totemPopManager)
    }

    @Mod.EventHandler
    fun postInit(event: FMLPostInitializationEvent) {
        logger.info("Loaded $NAME.")
        Display.setTitle(FULL)
    }
}